rust_src/src/lread.rs

```diff
@@ -404,4 +405,52 @@ pub fn read_char_exclusive(
     }
 }
 
+/// Delete the symbol named NAME, if any, from OBARRAY.
+/// The value is t if a symbol was found and deleted, nil otherwise.
+/// NAME may be a string or a symbol.  If it is a symbol, that symbol
+/// is deleted, if it belongs to OBARRAY--no other symbol is deleted.
+/// OBARRAY, if nil, defaults to the value of the variable `obarray'.
+/// usage: (unintern NAME OBARRAY)
+#[lisp_fn(min = "1")]
+pub fn unintern(name: LispSymbolOrString, obarray: Option<LispObarrayRef>) -> LispObject {
+    let obarray = obarray.unwrap_or_else(LispObarrayRef::global);
+    let obarray: LispObarrayRef = check_obarray(obarray.into()).into();
+
+    let tem = obarray.lookup(name);
+    if tem.is_integer() {
+        return Qnil;
+    }
+    // If arg was a symbol, don't delete anything but that symbol itself.
+    if name.is_symbol() && name != tem {
+        return Qnil;
+    }
+
+    let mut temp: LispSymbolRef = tem.into();
+    temp.set_uninterned();
+
+    let hash = unsafe { oblookup_last_bucket_number };
+
+    let mut obarray = LispObject::from(obarray).as_vector_or_error();
+    let symbol: LispSymbolRef = obarray.get(hash).into();
+
+    if symbol == temp {
+        match symbol.get_next() {
+            Some(sym) => obarray.set(hash, sym.into()),
+            None => obarray.set(hash, LispObject::from_natnum(0)),
+        };
+    } else {
```

Comment:
There should be a clear way to write this which embraces the iterators. Would you mind giving another go at it?

Reply:
Sure thing!

Reply:
Revised, should be cleaner now. The `if let ...` is probably unnecessary, but since the original C code silently passes even if the symbol is not found, I decided to leave it in there.